tests/unit/misc/userscripts/test_qute_lastpass.py

```diff
@@ -0,0 +1,353 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2014-2020 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Tests for misc.userscripts.qute-lastpass."""
+
+import os
+import json
+import subprocess
+from importlib.machinery import SourceFileLoader
+from importlib.util import spec_from_loader, module_from_spec
+from unittest.mock import MagicMock, call
+
+# qute-lastpass violates naming convention and does not have .py extension
+dirname = os.path.dirname(__file__)
+script_path = os.path.join(
+    dirname, "../../../../misc/userscripts/qute-lastpass")
+spec = spec_from_loader("qute_lastpass", SourceFileLoader(
+    "qute_lastpass",
+    script_path))
+qute_lastpass = module_from_spec(spec)
+spec.loader.exec_module(qute_lastpass)
+
+default_lpass_match = [
+    {
+        "id": "12345",
+        "name": "www.example.com",
+        "username": "fake@fake.com",
+        "password": "foobar",
+        "url": "https://www.example.com",
+    }
+]
+
+
+def get_response_mock(stdout='', stderr=''):
+    response = MagicMock()
+    response.stdout = stdout.encode()
+    response.stderr = stderr.encode()
+
+    return response
+
+
+def setup_subprocess_mock(mocker, stdout='', stderr=''):
+    mocker.patch('subprocess.run')
+
+    subprocess.run.return_value = get_response_mock(stdout, stderr)
+
+
+# Default arguments passed to qute-lastpass
+def get_arguments_mock(url):
+    arguments = MagicMock()
+    arguments.url = url
+    arguments.dmenu_invocation = 'rofi -dmenu'
+    arguments.insert_mode = True
+    arguments.io_encoding = 'UTF-8'
+    arguments.merge_candidates = False
+    arguments.password_only = False
+    arguments.username_only = False
+
+    return arguments
+
+
+class TestQuteLastPassComponents:
+    """Test qute-lastpass components."""
+
+    def test_fake_key_raw(self):
+        """Test if fake_key_raw properly escapes characters."""
+        qute_lastpass.qute_command = MagicMock()
```

Comment:
You should never patch things like this, because it'll never be reset and might affect other tests. Instead, this could be a pytest fixture which uses `mocker.patch`.

Also, please always use `spec=...` to avoid mocks allowing anything to be called on them. Here, you can probably use something like `spec=['__call__']` or so? Probably also should be a `Mock` and not a `MagicMock`.

Reply:
I actually tried to do it via fixture, but because of how `qute-lastpass` gets imported, it doesn't seem to work. I'll look into it a bit more, but I'm not sure if it's doable.

Reply:
Thinking about this further, this shouldn't affect other tests because of the way this module is imported (and why `mocker.patch('qute_lastpass.qute_command')` doesn't work), no?

```
repo_root = pathlib.Path(__file__).resolve().parents[4]
script_path = str(repo_root / 'misc' / 'userscripts' / 'qute-lastpass')
spec = spec_from_loader("qute_lastpass", SourceFileLoader(
    "qute_lastpass",
    script_path))
qute_lastpass = module_from_spec(spec)
spec.loader.exec_module(qute_lastpass)
```

If we are bypassing the normal `import` mechanism then we are bypassing caches as well. Inspecting `sys.modules` and `sys.path_importer_cache` will reveal no references to `qute-lastpass` or `qute_lastpass`.

`mocker.patch('qute_lastpass.qute_command')` seems to go through `__import__` so it will hit an exception and fail because it can't find `qute_lastpass`.

Therefore `qute_lastpass.qute_command = MagicMock()` should not be harmful because this is an override that is contained within this test. If another test were to adhoc import `qute-lastpass`, `module_from_spec` should be creating a separate `qute-lastpass` module.

I duplicated the test and removed the override to confirm that there's no shared mutation here:

```
tests/unit/misc/userscripts/test_qute_lastpass.py .............                                     [ 65%]
tests/unit/misc/userscripts/test_qute_lastpass2.py F

_______________________________ TestQuteLastPassMain2.test_main_happy_path ________________________________

self = <test_qute_lastpass2.TestQuteLastPassMain2 object at 0x7f3e94f392e0>
mocker = <pytest_mock.plugin.MockFixture object at 0x7f3e94fc1820>

    def test_main_happy_path(self, mocker):
        """Test sending username/password to qutebrowser on *single* match."""
        setup_subprocess_mock(mocker, json.dumps(default_lpass_match))

        arguments = get_arguments_mock(default_lpass_match[0]['url'])
>       exit_code = qute_lastpass.main(arguments)

tests/unit/misc/userscripts/test_qute_lastpass2.py:85:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
misc/userscripts/qute-lastpass:169: in main
    fake_key_raw(username)
misc/userscripts/qute-lastpass:110: in fake_key_raw
    qute_command('fake-key {}'.format(sequence))
misc/userscripts/qute-lastpass:80: in qute_command
    with open(os.environ['QUTE_FIFO'], 'w') as fifo:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = environ({'XDG_SESSION_TYPE': 'x11', 'XDG_VTNR': '7', 'XDG_SEAT': 'seat0', 'XDG_SESSION_DESKTOP': 'default', 'XDG_RUNTI...RRENT_TEST': 'tests/unit/misc/userscripts/test_qute_lastpass2.py::TestQuteLastPassMain2::test_main_happy_path (call)'})
key = 'QUTE_FIFO'

    def __getitem__(self, key):
        try:
            value = self._data[self.encodekey(key)]
        except KeyError:
            # raise KeyError with the original key value
>           raise KeyError(key) from None
E           KeyError: 'QUTE_FIFO'

/usr/lib/python3.8/os.py:675: KeyError
```

Reply:
It'd still be a problem for a second test in the same file - right now it might not be a problem in practice, but it's very unexpected to have those kind of side-effects when e.g. adding new tests in the future. You might be able to use `mocker.patch.object(qute_lastpass, `'qute_command'`, ...)` instead which does not (re)import anything.

Reply:
Yes, it looks like `mocker.patch.object` works. Thanks so much for your help.